spread/testflinger.go

```diff
@@ -0,0 +1,268 @@
+package spread
+
+import (
+	"fmt"
+	"gopkg.in/yaml.v2"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"golang.org/x/net/context"
+)
+
+func TestFlinger(p *Project, b *Backend, o *Options) Provider {
+	return &tfProvider{p, b, o}
+}
+
+type tfProvider struct {
+	project *Project
+	backend *Backend
+	options *Options
+}
+
+type tfServer struct {
+	p *tfProvider
+	d tfServerData
+
+	system  *System
+	address string
+}
+
+type tfServerData struct {
+	Name string
+	JobId string
+}
+
+type tfFile struct {
+    Queue          string `yaml:"job_queue"`
+    ProvisionDdata *tfProvisioningData `yaml:"provision_data"`
+    TestData       *tfTestData `yaml:"test_data"`
+    ReserveData    *tfReserveData `yaml:"reserve_data"`
+}
+
+var (
+    tfconfigfile = "testflinger.conf."
+    tfuser		 = "ubuntu"
+    tfpassword   = "ubuntu"
+)
+
+type tfReserveData struct {
+    SSHKeys []string `yaml:"ssh_keys"`
+}
+
+type tfTestData struct {
+    TestCommands []string `yaml:"test_cmds"`
+}
+
+type tfProvisioningData struct {
+    Url string
+}
+
+func (s *tfServer) String() string {
+	return fmt.Sprintf("%s (%s)", s.system, s.d.Name)
+}
+
+func (s *tfServer) Label() string {
+	return s.d.Name
+}
+
+func (s *tfServer) Provider() Provider {
+	return s.p
+}
+
+func (s *tfServer) Address() string {
+	return s.address
+}
+
+func (s *tfServer) System() *System {
+	return s.system
+}
+
+func (s *tfServer) ReuseData() interface{} {
+	return &s.d
+}
+
+func (s *tfServer) Discard(ctx context.Context) error {
+	output, err := exec.Command("testflinger-cli", "cancel", s.d.JobId).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("cannot discard testflinger reservation: %v", outputErr(output, err))
+	}
+	return nil
+}
+
+func (p *tfProvider) Backend() *Backend {
+	return p.backend
+}
+
+func (p *tfProvider) GarbageCollect() error {
+	return nil
+}
+
+func (p *tfProvider) Reuse(ctx context.Context, rsystem *ReuseSystem, system *System) (Server, error) {
+	s := &tfServer{
+		p:       p,
+		system:  system,
+		address: rsystem.Address,
+	}
+	err := rsystem.UnmarshalData(&s.d)
+	if err != nil {
+		return nil, fmt.Errorf("cannot unmarshal testflinger reuse data: %v", err)
+	}
+	return s, nil
+}
+
+func (p *tfProvider) Allocate(ctx context.Context, system *System) (Server, error) {
+	image := system.Image
+	name := tfName(system)
+
+	key := ""
+    if p.backend.Key != "" {
+		key = p.backend.Key
+	}
+
+    file := &tfFile{
+        Queue: name,
+        ProvisionDdata: &tfProvisioningData{
+            Url: image,
+        },
+        TestData: &tfTestData{
+            TestCommands: []string{"ssh $DEVICE_IP cat /proc/cpuinfo", "ssh $DEVICE_IP echo ubuntu:ubuntu | sudo chpasswd", "ssh $DEVICE_IP echo 'ubuntu ALL=(ALL) NOPASSWD:ALL' | sudo tee /etc/sudoers.d/create-user-ubuntu"},
```

Comment:
could you wrap this so we don't have to scroll horizontally as much? like
```go
TestCommands: []string{
	"ssh $DEVICE_IP cat /proc/cpuinfo",
	"ssh $DEVICE_IP echo ubuntu:ubuntu | sudo chpasswd",
	"ssh $DEVICE_IP echo 'ubuntu ALL=(ALL) NOPASSWD:ALL' | sudo tee /etc/sudoers.d/create-user-ubuntu",
},
```

Reply:
done